pvlib/modelchain.py

```diff
@@ -1410,10 +1680,46 @@ def run_model_from_effective_irradiance(self, data=None):
         pvlib.modelchain.ModelChain.run_model_from
         pvlib.modelchain.ModelChain.run_model_from_poa
         """
-
+        self._check_multiple_input(data)
         self._assign_weather(data)
         self._assign_total_irrad(data)
-        self.effective_irradiance = data['effective_irradiance']
+        self.results.effective_irradiance = _tuple_from_dfs(
+            data, 'effective_irradiance')
         self._run_from_effective_irrad(data)
 
         return self
+
+
+def _pairwise(iterable):
+    """s -> (s0,s1), (s1,s2), (s2, s3), ...
+
+    From the itertools cookbook.
+    """
+    a, b = itertools.tee(iterable)
+    next(b, None)
+    return zip(a, b)
+
+
+def _validate_indices(data):
+    indexes = map(lambda df: df.index, data)
+    if not all(itertools.starmap(pd.Index.equals,
+                                 _pairwise(indexes))):
+        raise ValueError("Input DataFrames must have same index.")
```

Comment:
Still overkill

```suggestion
def _validate_indices(data):
    indexes = map(lambda df: df.index, data)
    for index in indexes:
        if not index.equals(indexes[0]):
            raise ValueError("Input DataFrames must have same index.")
```

Also may want to rename.